### ISSUE TITLE
Graceful Shutdown Refactoring

### ISSUE THREAD

**Issue by ryanemerson** [infinispan/infinispan-operator#1154]
Graceful shutdown is used when a user sets `spec.replicas = 0` or when an upgrade is performed. The ultimate goal is to scale the cluster from `x` pods to 0, with each pod terminating normally so that all data is preserved when the pods are recreated. 

# Current Behaviour
1. A REST call is made to `/rest/v2/cluster?action=stop` on the first Infinispan pod marked as READY
2. Wait for all pods in the cluster to be NOT READY
3. Set the StatefulSet replicas to 0

## Limitations
- Once each Infinispan pod shutsdown in 1. they will always be restarted in order to satisfy the number of StatefulSet replicas
- It's possible that with a very large number of pods that one or more pods readinessProbe may pass and enter the READY state preventing conditions in stage 2 every being satisfied. Ultimately causing the cluster to not scale down.
- Verbose code required to handle steps 1-3.

# Proposal
Make use of StatefulSet semantics to scale down the cluster by setting `statefulSet.Spec.Replicas = 0`.

On Graceful shutdown request, do the following:

1. Disable cluster rebalancing via REST

This ensures that when scaling down from `{N-1..0}` cache entries aren't slowly migrated to pod 0.

2. Set `statefulSet.Spec.Replicas = 0`

The StatefulSet will start to scale down the cluster pods in the order `{N-1..0}`. Pods are shutdown gracefully as long as they don't exceed their configured `terminationGracePeriodSeconds` (30 seconds by default). We will need to experiment and ensure this is configured correctly.

3. On cluster scale up enable rebalancing (might not be necessary if this is config is not persisted, need to clarify).

# Missing
- Rebalancing can be disabled, however it is not yet exposed via the REST api.

**Comment by ryanemerson** [comment]
@rigazilla @dmvolod @Crumby These are just some thoughts after working on a upgrade issue today. Maybe I have missed some obvious limitations with this approach, please let me know what you think.

**Comment by dmvolod** [comment]
Very good proposal, @ryanemerson 
Once we are refactoring a graceful shutdown, it would be nice to skip step 1 in proposal for the ephemeral clusters, as graceful shutdown doesn't make sense for them

**Comment by dmvolod** [comment]
> The StatefulSet will start to scale down the cluster pods in the order `{N-1..0}`. Pods are shutdown gracefully as long as they don't exceed their configured `terminationGracePeriodSeconds` (30 seconds by default). We will need to experiment and ensure this is configured correctly.

We can try to use `Parallel Pod Management` in this case, but need to investigate more https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#pod-management-policies

**Comment by ryanemerson** [comment]
> We can try to use `Parallel Pod Management` in this case, but need to investigate more

I don't think parallel is a good fit for Infinispan as with this behaviour it's possible for clusters to scale down too quickly during a normal scaling operation, resulting in data loss. For example, consider a infinispan CR with `Replicas: 5` and caches with `num_owners=2`. If the user changes this to `Replicas: 2` and we have the parallel policy, it's possible that the pods containing the primary and backup of a cache entry could be removed from the StatefulSet before rebalancing completes. Utilising the `OrderedReady` strategy protects against this scenario, as the cluster can only be scaled down one pod at a time.

**Comment by rigazilla** [comment]
@ryanemerson looks good to me: much simpler and robust than the current one.
if possible I would keep the current implementation as a fall-back for a while, just to avoid problems with operator-operand compatibility.

**Comment by ryanemerson** [comment]
> if possible I would keep the current implementation as a fall-back for a while, just to avoid problems with operator-operand compatibility.

+1. There's two options here, both of which require exposing the ability to disable rebalance via REST in Infinispan 13.

1. Implement new GracefulShutdown behaviour AND keep old code for backwards compatibility with 12.1.x operators
2. Implement new behaviour for the second 13.x operator release, so that the require REST endpoint is available on both the old and new infinispan server during upgrades.

IMOk 2 will be much simpler and cleaner and there's no rush for this refactoring, so waiting shouldn't cause any issues.

**Comment by rigazilla** [comment]
I agree that 2 is better from the code POV, but I would evaluate 1 if the impacts on the code will be minimal. 
Mainly because 2 sets constraints on update: i.e. only operator 13.x ver.1 will support 12->13 migration and this forces the 12.x users to step into ver.1 even if a newer version is out.
Although this is handled automatically by OLM for OLM users, maybe we don't want to set this constraint for people that does installation via manual deployment.

**Comment by Crumby** [comment]
> this forces the 12.x users to step into ver.1 even if a newer version is out.

We are currently not allowing skipping versions during upgrades so it's probably not something to worry about unless we start introducing skipVersion in CSVs.

**Comment by ryanemerson** [comment]
> Mainly because 2 sets constraints on update: i.e. only operator 13.x ver.1 will support 12->13 migration and this forces the 12.x users to step into ver.1 even if a newer version is out.
> Although this is handled automatically by OLM for OLM users, maybe we don't want to set this constraint for people that does installation via manual deployment.

I think lots of our upgrade assumptions are already based upon this anyway, and AFAIK we only test the OLM scenario, so it's down to good fortune if a non-sequential upgrade path works atm.

**Comment by rigazilla** [comment]
> > Mainly because 2 sets constraints on update: i.e. only operator 13.x ver.1 will support 12->13 migration and this forces the 12.x users to step into ver.1 even if a newer version is out.
> > Although this is handled automatically by OLM for OLM users, maybe we don't want to set this constraint for people that does installation via manual deployment.
> 
> I think lots of our upgrade assumptions are already based upon this anyway, and AFAIK we only test the OLM scenario, so it's down to good fortune if a non-sequential upgrade path works atm.

Yeah I've exactly the same feeling, that's why I was proposing to avoid adding a new constraint if possible with little effort. 
Well when it comes down to discuss about upgrade things become blurry

**Comment by rigazilla** [comment]
> > this forces the 12.x users to step into ver.1 even if a newer version is out.
> 
> We are currently not allowing skipping versions during upgrades so it's probably not something to worry about unless we start introducing skipVersion in CSVs.

yeah this is how OLM works. I wondering what are the implications for a manual deploy since we're using a CD approach.

**Comment by ryanemerson** [comment]
PR to expose rebalancing configuration via REST: https://github.com/infinispan/infinispan/pull/9386

It will be available for 13.0.0.Final

**Comment by ryanemerson** [comment]
> yeah this is how OLM works. I wondering what are the implications for a manual deploy since we're using a CD approach.

Tbh I'm not sure how much people can/should rely on this for anything more than P.O.C deployments. Things often break in `master` introducing bugs etc, so I don't think we can really guarantee any upgrade stability across commits from this repo.

IMO support can only really be provided across tags and OLM deployments (upstream and downstream).

**Comment by rigazilla** [comment]
> Tbh I'm not sure how much people can/should rely on this for anything more than P.O.C deployments. Things often break in `master` introducing bugs etc, so I don't think we can really guarantee any upgrade stability across commits from this repo.
> 
> IMO support can only really be provided across tags and OLM deployments (upstream and downstream).

you're probably right @ryanemerson

**Comment by dmvolod** [comment]
We could identify set of available APIs on the each Infinispan cluster start and use sort of them depends of availability. It would be nice to have a OpenAPI/Swagger spec for that to reduce utilize a simple probes for that.

**Comment by ryanemerson** [comment]
Disabling rebalancing alone is not sufficient as we need shutdown to be called on the containers to ensure that data is correctly preserved. [ISPN-13271](https://issues.redhat.com/browse/ISPN-13271) adds a REST method that will shutdown all containers, but not terminate the server process. This will allow the operator execute the REST call and then set `spec.replicas = 0`.
